related_products_ecommerce/models/sale_order.py

```diff
@@ -0,0 +1,36 @@
+from odoo import api, models
+from odoo.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    def _cart_update(self, product_id, line_id=None, add_qty=1, set_qty=0, **kwargs):
+        """
+        - Users can modify the quantity of actual products (manually added products).
+        - Related products will automatically update their quantity based on the actual product's quantity.
+        - Users cannot manually change or remove related products from the cart.
+        """
+        res = super()._cart_update(product_id, line_id, add_qty, set_qty, **kwargs)
+        product = self.env["product.product"].browse(product_id)
+
+        if not product.exists():
+            raise ValidationError(self.env._("Product with ID %s not found") % product_id)
+
+        actual_product_line = self.order_line.filtered(lambda l: l.product_id.id == product_id)
+        actual_qty = actual_product_line.product_uom_qty if actual_product_line else 0
```

Comment:
```suggestion
        actual_qty = actual_product_line.product_uom_qty
```
```py
In [3]: self.env['sale.order.line'].product_uom_qty
Out[3]: 0.0
```
ORM returns 0.0 as it's float field on a empty recordset